File: engines/r-scripts/anova.R
# anova.R - One-Way ANOVA implementation for Method Studio
#
# Expected environment variables (set by wrapper.R from request data):
#   dependentVariable : list (length 1) of column name strings — the outcome variable slot
#   groupingVariable  : list (length 1) of column name strings — the grouping/factor slot
#   + columnar data with those column names present as bare variables
#
#   options$postHocTest      : "tukey" | "bonferroni" | "none"  (default "tukey")
#   options$confidenceLevel  : numeric confidence level, e.g. 0.95  (default 0.95)
#   options$effectSize       : logical — compute eta-squared (default TRUE)
#   options$missingValues    : "exclude-analysis" | "exclude-listwise"
#
# Returns a named list with:
#   f_statistic, df_between, df_within, p_value, significant,
#   eta_squared, eta_squared_label,
#   group_stats (list of per-group means, sds, ns),
#   post_hoc_results (Tukey HSD or Bonferroni comparisons, NULL if not run),
#   interpretation

if (!exists("dependentVariable") || length(dependentVariable) == 0) {
  stop("Variable slot 'dependentVariable' is required")
}
if (!exists("groupingVariable") || length(groupingVariable) == 0) {
  stop("Variable slot 'groupingVariable' is required")
}

# Variable slots come as lists from JSON — extract the column name strings
dep_name   <- if (is.list(dependentVariable)) dependentVariable[[1]] else dependentVariable[1]
group_name <- if (is.list(groupingVariable))  groupingVariable[[1]]  else groupingVariable[1]

if (!exists(dep_name))   stop(paste0("Column '", dep_name,   "' not found in data"))
if (!exists(group_name)) stop(paste0("Column '", group_name, "' not found in data"))

# Column data may come as lists from JSON — convert appropriately
dep_raw   <- get(dep_name)
group_raw <- get(group_name)

dependent <- as.numeric(if (is.list(dep_raw))   unlist(dep_raw)   else dep_raw)
group     <- if (is.list(group_raw)) unlist(group_raw) else group_raw

# ---- Read options ----

alpha    <- 0.05
post_hoc <- "tukey"

# Read flattened option variables directly
if (exists("confidenceLevel") && !is.null(confidenceLevel)) {
  alpha <- 1 - as.numeric(if (is.list(confidenceLevel)) confidenceLevel[[1]] else confidenceLevel)
}
if (exists("postHocTest") && !is.null(postHocTest)) {
  post_hoc <- as.character(if (is.list(postHocTest)) postHocTest[[1]] else postHocTest)
}

# ---- Validate and clean inputs ----

group <- as.factor(group)

keep      <- !is.na(dependent) & !is.na(group)
dependent <- dependent[keep]
group     <- droplevels(group[keep])

if (length(dependent) < 3) stop(paste0("'", dep_name, "' must have at least 3 non-missing values"))

n_groups <- nlevels(group)
if (n_groups < 2) stop(paste0("'", group_name, "' must have at least 2 distinct levels"))

group_levels <- levels(group)

# ---- Effect size label helper ----

eta_sq_label <- function(eta2) {
  if (eta2 < 0.01) {
    "negligible"
  } else if (eta2 < 0.06) {
    "small"
  } else if (eta2 < 0.14) {
    "medium"
  } else {
    "large"
  }
}

# ---- Run One-Way ANOVA ----

aov_fit   <- aov(dependent ~ group)
aov_table <- summary(aov_fit)[[1]]

ss_between <- aov_table["group",    "Sum Sq"]
ss_within  <- aov_table["Residuals","Sum Sq"]
ss_total   <- ss_between + ss_within

df_between <- aov_table["group",    "Df"]
df_within  <- aov_table["Residuals","Df"]

f_stat     <- aov_table["group",    "F value"]
p_val      <- aov_table["group",    "Pr(>F)"]

eta2       <- ss_between / ss_total

# ---- Per-group descriptives ----

group_stats <- lapply(group_levels, function(lv) {
  vals <- dependent[group == lv]
  list(
    n    = length(vals),
    mean = mean(vals),
    sd   = if (length(vals) > 1) sd(vals) else NA_real_,
    se   = if (length(vals) > 1) sd(vals) / sqrt(length(vals)) else NA_real_
  )
})
names(group_stats) <- group_levels

# ---- Post-hoc tests ----

post_hoc_results <- NULL

run_posthoc <- (post_hoc != "none") && (p_val < alpha)

if (run_posthoc) {
  if (post_hoc == "tukey" || post_hoc == "bonferroni") {
    ph     <- TukeyHSD(aov_fit, conf.level = 1 - alpha)
    ph_mat <- as.data.frame(ph$group)

    if (post_hoc == "bonferroni") {
      # Override Tukey p-values with Bonferroni-corrected ones
      raw_p <- pairwise.t.test(dependent, group,
                               p.adjust.method = "bonferroni")$p.value
      # Flatten the p-value matrix into comparison pairs
      pairs <- expand.grid(g1 = group_levels, g2 = group_levels,
                           stringsAsFactors = FALSE)
      pairs <- pairs[pairs$g1 < pairs$g2, ]

      bon_list <- lapply(seq_len(nrow(pairs)), function(i) {
        g1 <- pairs$g1[i]
        g2 <- pairs$g2[i]
        p_adj <- raw_p[max(g1, g2), min(g1, g2)]
        if (is.na(p_adj)) p_adj <- raw_p[min(g1, g2), max(g1, g2)]
        list(
          comparison  = paste0(g1, "-", g2),
          mean_diff   = mean(dependent[group == g1]) - mean(dependent[group == g2]),
          p_adjusted  = p_adj,
          significant = !is.na(p_adj) && p_adj < alpha,
          method      = "bonferroni"
        )
      })
      post_hoc_results <- bon_list
    } else {
      # Tukey HSD
      tukey_list <- lapply(rownames(ph_mat), function(comp) {
        list(
          comparison  = comp,
          mean_diff   = ph_mat[comp, "diff"],
          ci_lower    = ph_mat[comp, "lwr"],
          ci_upper    = ph_mat[comp, "upr"],
          p_adjusted  = ph_mat[comp, "p adj"],
          significant = ph_mat[comp, "p adj"] < alpha,
          method      = "tukey"
        )
      })
      post_hoc_results <- tukey_list
    }
  }
}

# ---- Visualisation ----

# Boxplot with jittered points
palette_cols <- c(
  "#4C72B0","#DD8452","#55A868","#C44E52",
  "#8172B2","#937860","#DA8BC3","#8C8C8C","#CCB974","#64B5CD"
)
group_cols <- palette_cols[seq_len(n_groups)]

boxplot(
  dependent ~ group,
  main   = sprintf("One-Way ANOVA: F(%d, %d) = %.3f, p %s %.4f",
                   df_between, df_within, f_stat,
                   if (p_val < alpha) "<" else ">=", alpha),
  xlab   = group_name,
  ylab   = dep_name,
  col    = group_cols,
  border = "#333333",
  notch  = FALSE
)
stripchart(
  dependent ~ group,
  vertical = TRUE,
  method   = "jitter",
  add      = TRUE,
  pch      = 20,
  col      = paste0(group_cols, "88")
)

# ---- Interpretation string ----

p_cmp <- if (p_val < alpha) "<" else ">="
sig_word <- if (p_val < alpha) "" else "not"
post_hoc_note <- if (!is.null(post_hoc_results)) {
  sprintf("Post-hoc %s tests were performed.", post_hoc)
} else if (p_val >= alpha) {
  "Post-hoc tests were not conducted (ANOVA not significant)."
} else {
  "Post-hoc tests were not requested."
}

interp <- sprintf(
  paste0(
    "One-Way ANOVA: F(%d, %d) = %.3f, p %s %.4f. ",
    "The effect of %s on %s is %s significant ",
    "(eta-squared = %.3f, %s effect). ",
    "%s"
  ),
  df_between,
  df_within,
  f_stat,
  p_cmp,
  alpha,
  group_name,
  dep_name,
  sig_word,
  eta2,
  eta_sq_label(eta2),
  post_hoc_note
)

# ---- Compose result ----

result <- list(
  f_statistic        = unname(f_stat),
  df_between         = df_between,
  df_within          = df_within,
  p_value            = p_val,
  significant        = p_val < alpha,
  eta_squared        = eta2,
  eta_squared_label  = eta_sq_label(eta2),
  group_stats        = group_stats,
  post_hoc_results   = post_hoc_results,
  post_hoc_method    = if (run_posthoc) post_hoc else "none",
  alpha              = alpha,
  interpretation     = interp
)

result

File: engines/python-scripts/anova.py
"""
ANOVA Script
============
Performs a One-Way Analysis of Variance (ANOVA).

Expected namespace variables (injected by engine):

  dependentVariable : list[str]
      Column name of the outcome / dependent variable (length 1).

  groupingVariable : list[str]
      Column name of the grouping / factor variable (length 1).

  data : dict
      Columnar data dictionary with column names as keys and arrays as values.

  -- Shared optional (via options dict) --
  options.postHocTest     : str     "tukey" | "bonferroni" | "none" (default "tukey").
  options.effectSize      : bool    Compute eta-squared (default True).
  options.confidenceLevel : float   Confidence level e.g. 0.95 → alpha = 0.05 (default 0.95).
  options.missingValues   : str     "exclude-analysis" | "exclude-listwise"

Result structure
----------------
{
  "f_statistic": float,
  "df_between": int,
  "df_within": int,
  "p_value": float,
  "significant": bool,
  "alpha": float,
  "effect_size": {
    "eta_squared": float,
    "interpretation": "negligible"|"small"|"medium"|"large"
  },
  "group_stats": {
    "<group_label>": {"n": int, "mean": float, "std": float, "se": float},
    ...
  },
  "post_hoc_results": [
    {
      "comparison": str,
      "mean_diff": float,
      "p_adjusted": float,
      "significant": bool,
      "ci_lower": float,   # Tukey only
      "ci_upper": float,   # Tukey only
      "method": "tukey" | "bonferroni"
    },
    ...
  ] | None,
  "post_hoc_method": "tukey" | "bonferroni" | "none",
  "interpretation": str
}
"""

from __future__ import annotations

import itertools
from typing import Any

import numpy as np
import pandas as pd
from scipy import stats as scipy_stats
from statsmodels.stats.multicomp import pairwise_tukeyhsd
from statsmodels.stats.multitest import multipletests


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _coerce(x) -> np.ndarray:
    """Convert list-like to a float64 array, dropping NaN."""
    arr = np.array(x, dtype=float)
    return arr[~np.isnan(arr)]


def _desc(arr: np.ndarray) -> dict:
    n = int(len(arr))
    mean = float(np.mean(arr))
    std = float(np.std(arr, ddof=1)) if n > 1 else 0.0
    se = std / np.sqrt(n) if n > 0 else 0.0
    return {
        "n": n,
        "mean": round(mean, 6),
        "std": round(std, 6),
        "se": round(se, 6),
    }


def _eta_sq_label(eta2: float) -> str:
    if eta2 < 0.01:
        return "negligible"
    elif eta2 < 0.06:
        return "small"
    elif eta2 < 0.14:
        return "medium"
    else:
        return "large"


# ---------------------------------------------------------------------------
# Resolve inputs from injected namespace
# ---------------------------------------------------------------------------

# Variable slots come as lists of column name strings
if "dependentVariable" not in dir() or len(dependentVariable) == 0:  # noqa: F821
    raise ValueError("Variable slot 'dependentVariable' is required")
if "groupingVariable" not in dir() or len(groupingVariable) == 0:  # noqa: F821
    raise ValueError("Variable slot 'groupingVariable' is required")

dep_name   = dependentVariable[0]   # noqa: F821
group_name = groupingVariable[0]    # noqa: F821

# Access column data from the data dictionary
_data = data if "data" in dir() else {}  # noqa: F821

if dep_name not in _data:
    raise ValueError(f"Column '{dep_name}' not found in data")
if group_name not in _data:
    raise ValueError(f"Grouping column '{group_name}' not found in data")

# Read options
_options = options if "options" in dir() else {}  # noqa: F821
_confidence_level = float(_options.get("confidenceLevel", 0.95)) if _options else 0.95
_alpha: float = round(1.0 - _confidence_level, 10)
_post_hoc: str = str(_options.get("postHocTest", "tukey")).lower() if _options else "tukey"

# Build a DataFrame from the raw column data for aligned NA removal
_df = pd.DataFrame({
    "y": list(_data[dep_name]),
    "g": [str(v) for v in _data[group_name]],
})
_df["y"] = pd.to_numeric(_df["y"], errors="coerce")
_df = _df.dropna(subset=["y", "g"])
_df["g"] = _df["g"].astype(str)

_dep_arr = _df["y"].to_numpy(dtype=float)
_grp_arr = _df["g"].to_numpy()

_group_labels = sorted(_df["g"].unique().tolist())
_n_groups = len(_group_labels)

if len(_dep_arr) < 3:
    raise ValueError(f"'{dep_name}' must have at least 3 non-missing values")
if _n_groups < 2:
    raise ValueError(f"'{group_name}' must have at least 2 distinct levels")

# ---------------------------------------------------------------------------
# Split into per-group arrays
# ---------------------------------------------------------------------------

_groups_data: dict[str, np.ndarray] = {
    lv: _dep_arr[_grp_arr == lv] for lv in _group_labels
}

# ---------------------------------------------------------------------------
# One-Way ANOVA via scipy (fast, minimal deps)
# ---------------------------------------------------------------------------

_f_stat, _p_val = scipy_stats.f_oneway(*_groups_data.values())
_f_stat = float(_f_stat)
_p_val = float(_p_val)

_df_between = int(_n_groups - 1)
_df_within = int(len(_dep_arr) - _n_groups)

# ---------------------------------------------------------------------------
# Eta-squared
# ---------------------------------------------------------------------------

_grand_mean = float(np.mean(_dep_arr))
_ss_between = float(sum(
    len(arr) * (float(np.mean(arr)) - _grand_mean) ** 2
    for arr in _groups_data.values()
))
_ss_total = float(np.sum((_dep_arr - _grand_mean) ** 2))
_eta2 = _ss_between / _ss_total if _ss_total > 0 else 0.0

# ---------------------------------------------------------------------------
# Per-group descriptives
# ---------------------------------------------------------------------------

_group_stats: dict[str, Any] = {lv: _desc(arr) for lv, arr in _groups_data.items()}

# ---------------------------------------------------------------------------
# Post-hoc tests (only when ANOVA is significant)
# ---------------------------------------------------------------------------

_post_hoc_results = None
_run_posthoc = (_post_hoc != "none") and (_p_val < _alpha)

if _run_posthoc:
    if _post_hoc == "tukey":
        _tukey = pairwise_tukeyhsd(_dep_arr, _grp_arr, alpha=_alpha)
        _tukey_df = pd.DataFrame(
            data=_tukey.summary().data[1:],
            columns=_tukey.summary().data[0],
        )
        _post_hoc_results = []
        for _, row in _tukey_df.iterrows():
            _post_hoc_results.append({
                "comparison": f"{row['group1']}-{row['group2']}",
                "mean_diff": round(float(row["meandiff"]), 6),
                "ci_lower": round(float(row["lower"]), 6),
                "ci_upper": round(float(row["upper"]), 6),
                "p_adjusted": round(float(row["p-adj"]), 8),
                "significant": bool(row["reject"]),
                "method": "tukey",
            })

    elif _post_hoc == "bonferroni":
        _pairs = list(itertools.combinations(_group_labels, 2))
        _raw_p_vals = []
        _pair_info = []
        for g1, g2 in _pairs:
            arr1 = _groups_data[g1]
            arr2 = _groups_data[g2]
            _, raw_p = scipy_stats.ttest_ind(arr1, arr2, equal_var=False)
            _raw_p_vals.append(float(raw_p))
            _pair_info.append((g1, g2, float(np.mean(arr1)) - float(np.mean(arr2))))

        _reject, _adj_p, _, _ = multipletests(_raw_p_vals, alpha=_alpha, method="bonferroni")
        _post_hoc_results = []
        for (g1, g2, diff), p_adj, sig in zip(_pair_info, _adj_p, _reject):
            _post_hoc_results.append({
                "comparison": f"{g1}-{g2}",
                "mean_diff": round(diff, 6),
                "p_adjusted": round(float(p_adj), 8),
                "significant": bool(sig),
                "method": "bonferroni",
            })

# ---------------------------------------------------------------------------
# Interpretation string
# ---------------------------------------------------------------------------

_sig_str = "statistically significant" if _p_val < _alpha else "not statistically significant"
_interp = (
    f"One-Way ANOVA: F({_df_between}, {_df_within}) = {_f_stat:.3f}, "
    f"p {'<' if _p_val < _alpha else '>='} {_alpha:.2f}. "
    f"The effect of {group_name} on {dep_name} is {_sig_str} "
    f"(eta-squared = {_eta2:.3f}, {_eta_sq_label(_eta2)} effect). "
)
if _post_hoc_results is not None:
    _interp += f"Post-hoc {_post_hoc} tests were performed."
elif _p_val >= _alpha:
    _interp += "Post-hoc tests were not conducted (ANOVA not significant)."
else:
    _interp += "Post-hoc tests were not requested."

# ---------------------------------------------------------------------------
# Compose result
# ---------------------------------------------------------------------------

result = {
    "f_statistic": round(_f_stat, 6),
    "df_between": _df_between,
    "df_within": _df_within,
    "p_value": round(_p_val, 8),
    "significant": bool(_p_val < _alpha),
    "alpha": _alpha,
    "effect_size": {
        "eta_squared": round(_eta2, 6),
        "interpretation": _eta_sq_label(_eta2),
    },
    "group_stats": _group_stats,
    "post_hoc_results": _post_hoc_results,
    "post_hoc_method": _post_hoc if _run_posthoc else "none",
    "interpretation": _interp,
}
